kernel/proc.c

```diff
@@ -567,8 +620,10 @@ void
 wakeup(void *chan)
 {
   struct proc *p;
+  struct proc_list* cur_proc;
 
-  for(p = proc; p < &proc[NPROC]; p++) {
+  for(cur_proc = proc_list->next; cur_proc != proc_list; cur_proc = cur_proc->next) {
```

Comment:
race condition on proc_list during iteration?

Reply:
I mean that you can change cur->next between for-loop condition and cur = cur->next.

Reply:
Not really because I always hold proc_list_lock while in scheduler and I need to hold this lock to remove something. So elements are removed only in between of scheduler loops and can't affect them. My allocproc also can't affect next of any proc but head and I hold head->lock during adding new element to list so there are no potential races.

Reply:
AAAA!!! OK, UNDERSTAND! This is brilliant words, my friend